### PR TITLE
BREAKING CHANGE: DecodeLimit and DecodeAll extensions now advance input

### LINKED ISSUE #287: Merge `decode_with_depth_limit` with `decode_and_advance_with_depth_limit`
#286 added a `decode_and_advance_with_depth_limit` associated method for the `DepthLimit` trait to advance the input pointer while decoding an item. We would like to merge this with `decode_with_depth_limit` once we release a new major version.

### LINKED ISSUE #313: `decode_all` should consume the input
`decode_all` should consume the `input` to be more consistent with the normal `Decode` interface. By consuming I mean `&mut &[u8]`, so that we advance the input as we decode it.

### PR DESCRIPTION
**Pull request by wigy-opensource-developer**
Fixes #287
Fixes #313

### CODE REVIEW

**Review by KiChjang — DISMISSED** [review]
Whitespace!!!! We use tabs instead of spaces here

**Comment by wigy-opensource-developer** [comment]
> Whitespace!!!! We use tabs instead of spaces here

Oh. Checking my IDE and `cargo fmt` settings... Maybe it is different in Substrate and Polkadot.

File: src/decode_all.rs [hunk 1]
```diff
@@ -23,15 +23,14 @@ pub trait DecodeAll: Sized {
 	/// Decode `Self` and consume all of the given input data.
 	///
 	/// If not all data is consumed, an error is returned.
-	fn decode_all(input: &[u8]) -> Result<Self, Error>;
+	fn decode_all<I: Input>(input: &mut I) -> Result<Self, Error>;
```
Comment (bkchr) [inline src/decode_all.rs:26]: Please just keep `[u8]`, because `Input` doesn't give you any guarantee that it can return a `remaining_len`

Comment (bkchr) [reply src/decode_all.rs:26]: Or we document this. But then we also need to return the error of `remaining_len`. If remaining_len would return `None` we would need to try to read on byte to ensure that there isn't anything anymore.

**Comment by bkchr** [comment]
> > Whitespace!!!! We use tabs instead of spaces here
> 
> Oh. Checking my IDE and `cargo fmt` settings... Maybe it is different in Substrate and Polkadot.

You probably need to copy the rustfmt.toml file. You can add it here in the pr.

**Comment by wigy-opensource-developer** [comment]
This has been a rollercoaster, but hopefully everything is better than how it started.

**Review by bkchr — APPROVED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "parity-scale-codec"
 description = "SCALE - Simple Concatenating Aggregated Little Endians"
-version = "2.3.1"
+version = "3.0.0"
```
Comment (bkchr) [inline Cargo.toml:4]: We should also bump the proc-macro.

File: .rustfmt.toml [hunk 3]
```diff
@@ -0,0 +1 @@
+hard_tabs = true
```
Comment (bkchr) [inline .rustfmt.toml:1]: Please just copy it from Substrate

**Comment by wigy-opensource-developer** [comment]
bot merge

**Comment by wigy-opensource-developer** [comment]
@KiChjang Could you please have a second look on this? I would like to merge to master and release 3.0.0

**Review by KiChjang — APPROVED** [review]

**Pull request merged**